Assets/Documentation/man/man3/TuerRESET.3
.TH "TuerRESET" 3 "Version 0.1" "NPP VR Simulation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
TuerRESET \- This class implements logic to reset the simulation by using the door handle in the scene of the VR application\&.  

.SH SYNOPSIS
.br
.PP
.PP
Inherits MonoBehaviour\&.
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "InputActionReference \fBtrigger\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fItrigger\fP is a reference to an input action
.RE
.PP
"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBStart\fP ()"
.br
.RI "This method initialises the InputActionReference trigger\&. "
.ti -1c
.RI "void \fBUpdate\fP ()"
.br
.RI "This method checks whether the door handle is being hovered over and whether the trigger is being pressed\&. "
.ti -1c
.RI "void \fBOnEnable\fP ()"
.br
.RI "This method is called when the object is enabled and adds event listeners for the selectEntered and selectExited events\&. "
.ti -1c
.RI "void \fBOnDisable\fP ()"
.br
.RI "This method is called when the object is disabled and removes event listeners for the selectEntered and selectExited events\&. "
.ti -1c
.RI "void \fBOnHoverEntered\fP (HoverEnterEventArgs args)"
.br
.RI "This method is called when the XR interactor looks at the object, i\&.e\&. "
.ti -1c
.RI "void \fBOnHoverExited\fP (HoverExitEventArgs args)"
.br
.RI "This method is called when the XR interactor is no longer looking at the object, i\&.e\&. "
.ti -1c
.RI "IEnumerator \fBHandleDoorInteraction\fP ()"
.br
.RI "This method is called when the player interacts with the door handle via the XR interactor to reset the simulation\&. "
.ti -1c
.RI "IEnumerator \fBRotateToAngle\fP (float targetAngle, float duration)"
.br
.RI "This method is called in \fBHandleDoorInteraction()\fP to rotate the door handle\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "bool \fBisCooldown\fP = false"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIisCooldown\fP checks if the handle has been used recently 
.RE
.PP
"
.ti -1c
.RI "bool \fBisHovering\fP = false"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIisHovering\fP checks if the XR interactor is hovering over the handle 
.RE
.PP
"
.ti -1c
.RI "\fBNPPClient\fP \fBnppClient\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fInppClient\fP is a reference to the \fBNPPClient\fP instance in the scene
.RE
.PP
"
.in -1c
.SH "Detailed Description"
.PP 
This class implements logic to reset the simulation by using the door handle in the scene of the VR application\&. 
.SH "Member Function Documentation"
.PP 
.SS "IEnumerator TuerRESET\&.HandleDoorInteraction ()\fR [private]\fP"

.PP
This method is called when the player interacts with the door handle via the XR interactor to reset the simulation\&. 
.SS "void TuerRESET\&.OnDisable ()\fR [private]\fP"

.PP
This method is called when the object is disabled and removes event listeners for the selectEntered and selectExited events\&. 
.SS "void TuerRESET\&.OnEnable ()\fR [private]\fP"

.PP
This method is called when the object is enabled and adds event listeners for the selectEntered and selectExited events\&. 
.SS "void TuerRESET\&.OnHoverEntered (HoverEnterEventArgs args)\fR [private]\fP"

.PP
This method is called when the XR interactor looks at the object, i\&.e\&. points to or moves onto the object \&.

.PP
\fBParameters\fP
.RS 4
\fIargs\fP passes event specific arguments upon entering the interaction
.RE
.PP

.SS "void TuerRESET\&.OnHoverExited (HoverExitEventArgs args)\fR [private]\fP"

.PP
This method is called when the XR interactor is no longer looking at the object, i\&.e\&. the interactor moves away\&.

.PP
\fBParameters\fP
.RS 4
\fIargs\fP passes event specific arguments upon exiting the interaction
.RE
.PP

.SS "IEnumerator TuerRESET\&.RotateToAngle (float targetAngle, float duration)\fR [private]\fP"

.PP
This method is called in \fBHandleDoorInteraction()\fP to rotate the door handle\&. 
.PP
\fBParameters\fP
.RS 4
\fItargetAngle\fP specifies the angle to rotate the door handle to
.br
\fIduration\fP specifies the duration of the rotation
.RE
.PP

.SS "void TuerRESET\&.Start ()\fR [private]\fP"

.PP
This method initialises the InputActionReference trigger\&. 
.SS "void TuerRESET\&.Update ()\fR [private]\fP"

.PP
This method checks whether the door handle is being hovered over and whether the trigger is being pressed\&. 
.SH "Member Data Documentation"
.PP 
.SS "bool TuerRESET\&.isCooldown = false\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fIisCooldown\fP checks if the handle has been used recently 
.RE
.PP

.SS "bool TuerRESET\&.isHovering = false\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fIisHovering\fP checks if the XR interactor is hovering over the handle 
.RE
.PP

.SS "\fBNPPClient\fP TuerRESET\&.nppClient\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fInppClient\fP is a reference to the \fBNPPClient\fP instance in the scene
.RE
.PP

.SS "InputActionReference TuerRESET\&.trigger"

.PP

.PP
\fBParameters\fP
.RS 4
\fItrigger\fP is a reference to an input action
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for NPP VR Simulation from the source code\&.


Assets/Skripte/TuerRESET.cs
using System;
using UnityEngine;
using UnityEngine.XR.Interaction.Toolkit;
using System.Collections;
using System.Collections.Generic;
using UnityEngine.InputSystem;
using UnityEngine.XR;
using UnityEngine.XR.Interaction.Toolkit.Inputs;
using UnityEngine.XR.Interaction.Toolkit.Inputs.Readers;
using UnityEngine.XR.Interaction.Toolkit.Interactables;

/// <summary>
/// This class implements logic to reset the simulation by using the door handle in the scene of the VR application.
/// </summary>
public class TuerRESET : MonoBehaviour
{



    /// <param name="isCooldown"> checks if the handle has been used recently </param>
    private bool isCooldown = false;
    /// <param name="isHovering"> checks if the XR interactor is hovering over the handle </param>    
    private bool isHovering = false;
    /// <param name="trigger"> is a reference to an input action</param>
    public InputActionReference trigger;
    /// <param name="nppClient"> is a reference to the NPPClient instance in the scene</param>
    NPPClient nppClient;   


    /// <summary>
    /// This method initialises the InputActionReference trigger.
    /// </summary>
    private void Start()
    {
        nppClient = FindObjectOfType<NPPClient>();
        InputActionManager temp = FindObjectOfType<InputActionManager>();
        if (temp != null && temp.actionAssets.Count > 0)
        {
            foreach (var act in temp.actionAssets )
            {
                var action = act.FindAction("XRI Right Interaction/Select");
                trigger = InputActionReference.Create(action);
                break;
            }
        }
    }

    /// <summary>
    /// This method checks whether the door handle is being hovered over and whether the trigger is being pressed.
    /// </summary>
    private void Update()
    {
        if (isHovering && !isCooldown && trigger.action.triggered)
        {
            StartCoroutine(HandleDoorInteraction());
        }
    }

    /// <summary>
    /// This method is called when the object is enabled and adds event listeners for the selectEntered and selectExited events.
    /// </summary>
    private void OnEnable()
    {
        var interactable = GetComponent<XRSimpleInteractable>();
        interactable.hoverEntered.AddListener(OnHoverEntered);
        interactable.hoverExited.AddListener(OnHoverExited);
    }

    /// <summary>
    /// This method is called when the object is disabled and removes event listeners for the selectEntered and selectExited events.
    /// </summary>
    private void OnDisable()
    {
        var interactable = GetComponent<XRSimpleInteractable>();
        interactable.hoverEntered.RemoveListener(OnHoverEntered);
        interactable.hoverExited.RemoveListener(OnHoverExited);
    }

    /// <summary>
    /// This method is called when the XR interactor looks at the object, i.e. points to or moves onto the object .
    /// </summary>
    /// <param name="args"> passes event specific arguments upon entering the interaction</param>
    private void OnHoverEntered(HoverEnterEventArgs args)
    {
        isHovering = true;
    }

    /// <summary>
    /// This method is called when the XR interactor is no longer looking at the object, i.e. the interactor moves away.
    /// </summary>
    /// <param name="args"> passes event specific arguments upon exiting the interaction</param>
    private void OnHoverExited(HoverExitEventArgs args)
    {
        isHovering = false;
    }
    
    /// <summary>
    /// This method is called when the player interacts with the door handle via the XR interactor to reset the simulation.
    /// </summary>
    private IEnumerator HandleDoorInteraction()
    {
        isCooldown = true;
        // Reset the simulation
        StartCoroutine(nppClient.ResetSimulation());

        FindAnyObjectByType<AusfallAnzeigenManager>().SetAllLampsToWhite();
        
        // Smoothly rotate to 40 degrees on the Z axis over 0.5 seconds
        yield return RotateToAngle(40, 0.35f);

        // Smoothly rotate back to 0 degrees on the Z axis over 0.5 seconds
        yield return RotateToAngle(0, 0.35f);

        // Wait for 1 second before allowing another interaction
        yield return new WaitForSeconds(1f);

        isCooldown = false;

        // RESET
    }

    /// <summary>
    /// This method is called in HandleDoorInteraction() to rotate the door handle.
    /// </summary>
    /// <param name="targetAngle">  specifies the angle to rotate the door handle to</param>
    /// <param name="duration"> specifies the duration of the rotation</param>
    private IEnumerator RotateToAngle(float targetAngle, float duration)
    {
        float startAngle = transform.localRotation.eulerAngles.z;
        float time = 0;

        while (time < duration)
        {
            float angle = Mathf.Lerp(startAngle, targetAngle, time / duration);
            transform.localRotation = Quaternion.Euler(0, 0, angle);
            time += Time.deltaTime;
            yield return null;
        }

        transform.localRotation = Quaternion.Euler(0, 0, targetAngle);
    }
}